### PR TITLE
WIP: JS API: Fixes & maintenance: commands new bases

### PR DESCRIPTION
**Pull request by iNewLegend**
Moved: https://github.com/elementor/elementor/pull/12082

### CODE REVIEW

**Comment by iNewLegend** [comment]
TODO: Add async hooks to `commands.js`

File: assets/dev/js/editor/document/commands/base/command-history.js [hunk 1]
```diff
@@ -1,13 +1,11 @@
-import CommandBase from 'elementor-api/modules/command-base';
+import Command from 'elementor-api/modules/command';
 
-export default class CommandHistory extends CommandBase {
+export default class CommandHistory extends Command {
 	static getInstanceType() {
 		return 'CommandHistory';
 	}
 
-	constructor( args ) {
```
Comment (matipojo) [inline assets/dev/js/editor/document/commands/base/command-history.js:8]: Please separate the fixes from the renaming. keep this PR for rename only.

File: core/common/assets/js/api/core/commands.js [hunk 2]
```diff
@@ -1,4 +1,8 @@
+import CommandBase from 'elementor-api/modules/command-base';
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:1]: It shouldn't be `command`?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/common/assets/js/api/core/commands.js (hunk 2 above)
Comment (iNewLegend) [reply core/common/assets/js/api/core/commands.js]: **CommandBase** is the base class parent of all commands, later we validate if the command is instance of **CommandBase**.

File: core/common/assets/js/api/modules/command-callback.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	static getInstanceType() {
+		return 'CommandCallback';
+	}
+
+	static getCallback() {
+		return this.registerArgs.__callback;
+	}
+
+	apply( args = {} ) {
+		return this.constructor.getCallback().apply( $e.commands.getComponent( $e.commands.getCurrentLast() ), [ args ] );
```
Comment (iNewLegend) [inline core/common/assets/js/api/modules/command-callback.js:13]: use `this.constructor.getComponent()`

File: core/common/assets/js/api/core/commands.js [hunk 4]
```diff
@@ -1,31 +1,32 @@
...
+	constructor( ... args ) {
+		super( ... args );
 
 		this.current = {};
 		this.currentArgs = {};
 		this.currentTrace = [];
 		this.commands = {};
 		this.components = {};
-
 		this.classes = {};
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:22]: delete.

File: core/common/assets/js/api/core/commands.js [hunk 5]
```diff
@@ -269,32 +284,76 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+				return false;
+			}
 		}
 
-		this.afterRun( command );
+		const onAfter = ( _results ) => {
+			// For data hooks.
+			instance.onAfterApply( instance.args, _results );
+
+			if ( instance.isDataChanged() ) {
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:337]: should be part of *CommandEditor*.

File: core/common/assets/js/api/core/commands.js [hunk 6]
```diff
@@ -269,32 +284,76 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+			}
+
+			// For UI hooks.
+			instance.onAfterRun( instance.args, _results );
 
-		this.trigger( 'run:after', component, command, args, results );
+			this.afterRun( command, args, results );
+		};
+
+		// TODO: Temp code determine if it's a jQuery deferred object.
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:347]: Remove JQuery defrered.

File: core/common/assets/js/api/core/commands.js [hunk 7]
```diff
@@ -269,32 +284,76 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+			return results;
 		}
 
-		const results = this.commands[ command ].apply( component, [ args ] );
+		// TODO: Check with mati.
+		if ( ! this.validateInstance( instance, command ) ) {
+			return;
+		}
 
-		// TODO: Consider add results to `$e.devTools`.
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:290]: add to devtools tasks

File: assets/dev/js/editor/document/base/command-disable-enable.js [hunk 8]
```diff
@@ -22,16 +22,16 @@ export default class CommandDisableEnable extends CommandHistory {
 
 	static restore( historyItem, isRedo ) {
 		const data = historyItem.get( 'data' ),
-			CommandClass = $e.commands.getCommandClass( data.command );
```
Comment (iNewLegend) [inline assets/dev/js/editor/document/base/command-disable-enable.js:25]: UPPERCASE

File: core/common/assets/js/api/core/commands.js [hunk 9]
```diff
@@ -1,31 +1,31 @@
...
-
-		this.classes = {};
 	}
 
 	/**
-	 * @param id
-	 * @returns {CommandBase}
+	 * @returns {Command}
 	 */
 	getCommandClass( id ) {
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:27]: Can be deleted.

**Pull request closed without merging**